README.md
ARTool: R Package for the Aligned Rank Transform for Nonparametric Factorial ANOVAs
===================================================================================

[![Build Status](https://travis-ci.org/mjskay/ARTool.png?branch=master)](https://travis-ci.org/mjskay/ARTool) [![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/ARTool)](http://cran.r-project.org/web/packages/ARTool) [![GPL &gt;= 2](https://img.shields.io/badge/GPL-%E2%89%A52-brightgreen.svg)](https://cran.r-project.org/web/licenses/GPL-3) [![DOI](https://zenodo.org/badge/19809/mjskay/ARTool.svg)](https://zenodo.org/badge/latestdoi/19809/mjskay/ARTool) [![DOI](https://img.shields.io/badge/DOI-10.1145%2F1978942.1978963-blue.svg)](http://dx.doi.org/10.1145/1978942.1978963)

*Matthew Kay, University of Washington <mjskay@umich.edu>*<br> *Jacob O. Wobbrock, University of Washington <wobbrock@uw.edu>*

ARTool is an R package implementing the Aligned Rank Transform for conducting nonparametric analyses of variance on factorial models. This implementation is based on the ART procedure as used in the original implementation of [ARTool](http://depts.washington.edu/aimgroup/proj/art/) by Wobbrock et al.

The package automates the Aligning-and-Ranking process using the `art` function. It also automates the process of running a series of ANOVAs on the transformed data and extracting the results of interest. It supports traditional ANOVA models (fit using `lm`), repeated measures ANOVAs (fit using `aov`), and mixed effects models (fit using `lmer`); the model used is determined by the formula passed to `art`.

**Note**: The documentation of this package assumes some level of familiarity with when and why you may want to use the aligned rank transform; the [ARTool page](http://depts.washington.edu/aimgroup/proj/art/) provides a more in-depth (and highly approachable) introduction to the aligned rank transform and the motivation for its use.

Installation
------------

You can install the latest released version from CRAN with this R command:

``` r
install.packages("ARTool")
```

**Or**, you can install the latest development version from GitHub with these R commands:

``` r
install.packages("devtools")
devtools::install_github("mjskay/ARTool")
```

Example
-------

The general approach to using ART is to transform your data using `art` , verify the ART procedure is appropriate to the dataset using `summary` , and then run an anova on the transformed data using `anova` .

First, let us load some example data:

``` r
library(ARTool)
data(Higgins1990Table5)
```

`Higgins1990Table5` is a data frame from an experiment in which the effects of `Moisture` and `Fertilizer` on `DryMatter` in peat pots was tested. Four pots were placed on each `Tray` , with `Moisture` varied between `Tray` s and `Fertilizer` varied within `Tray` s. We can see the basic structure of the data:

``` r
str(Higgins1990Table5)
```

    ## 'data.frame':    48 obs. of  4 variables:
    ##  $ Tray      : Factor w/ 12 levels "t1","t2","t3",..: 1 1 1 1 2 2 2 2 3 3 ...
    ##  $ Moisture  : Factor w/ 4 levels "m1","m2","m3",..: 1 1 1 1 1 1 1 1 1 1 ...
    ##  $ Fertilizer: Factor w/ 4 levels "f1","f2","f3",..: 1 2 3 4 1 2 3 4 1 2 ...
    ##  $ DryMatter : num  3.3 4.3 4.5 5.8 4 4.1 6.5 7.3 1.9 3.8 ...

``` r
head(Higgins1990Table5, n=8)
```

    ##   Tray Moisture Fertilizer DryMatter
    ## 1   t1       m1         f1       3.3
    ## 2   t1       m1         f2       4.3
    ## 3   t1       m1         f3       4.5
    ## 4   t1       m1         f4       5.8
    ## 5   t2       m1         f1       4.0
    ## 6   t2       m1         f2       4.1
    ## 7   t2       m1         f3       6.5
    ## 8   t2       m1         f4       7.3

### Step 1: Transform the data

To analyze this data using the aligned rank transform, we first transform the data using `art` . We specify the response variable (`DryMatter` ), the fixed effects and all of their interactions (`Moisture*Fertilizer`, or equivalently `Moisture + Fertilizer + Moisture:Fertilizer`), and any grouping terms if present (here, `(1|Tray)` ).

While `(1|Tray)` has no effect on the results of the aligned rank transformation, it will be used by `anova` to determine the type of model to run: when grouping terms are present, mixed effects models are run using `lmer`. If you wish to use a repeated measures ANOVA instead of a mixed effects model, you can use an `Error` term instead (see below for an example of this). If you do not having repeated measures, do not include any grouping terms or error terms.

``` r
m <- art(DryMatter ~ Moisture*Fertilizer + (1|Tray), data=Higgins1990Table5)
```

### Step 2: Verify appropriateness of ART

To verify that the ART procedure was correctly applied and is appropriate for this dataset, we can look at the output of `summary` :

``` r
summary(m)
```

    ## Aligned Rank Transform of Factorial Model
    ## 
    ## Call:
    ## art(formula = DryMatter ~ Moisture * Fertilizer + (1 | Tray), 
    ##     data = Higgins1990Table5)
    ## 
    ## Column sums of aligned responses (should all be ~0):
    ##            Moisture          Fertilizer Moisture:Fertilizer 
    ##                   0                   0                   0 
    ## 
    ## F values of ANOVAs on aligned responses not of interest (should all be ~0):
    ##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
    ##       0       0       0       0       0       0

We see that the columns sums of aligned responses and the F values of ANOVAs on aligned responses not of interest are all ~0, indicating that the alignment correctly "stripped out" effects not of interest. Thus, we can apply the ANOVA on the transformed data.

### Step 3: Run the ANOVA

ARTool automatically selects the model to be used for the ANOVA. Because we have included a grouping term, `(1|Tray)`, ARTool will fit mixed effects models using `lmer` and run the ANOVAs on them:

``` r
anova(m)
```

    ## Analysis of Variance of Aligned Rank Transformed Data
    ## 
    ## Table Type: Analysis of Deviance Table (Type III Wald F tests with Kenward-Roger df) 
    ## Model: Mixed Effects (lmer)
    ## Response: art(DryMatter)
    ## 
    ##                             F Df Df.res     Pr(>F)    
    ## 1 Moisture             23.833  3      8 0.00024199 ***
    ## 2 Fertilizer          122.402  3     24 1.1124e-14 ***
    ## 3 Moisture:Fertilizer   5.118  9     24 0.00064665 ***
    ## ---
    ## Signif. codes:   0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

### Alternative model: Repeated Measures ANOVA

This particular study could also be analyzed using a repeated measures ANOVA, yielding the same results (note that repeated measures ANOVAs and mixed effects models will not always yield the same results). To instead run a repeated measures ANOVA, add an `Error` term to the model as you might for a call to `aov`:

``` r
m <- art(DryMatter ~ Moisture*Fertilizer + Error(Tray), data=Higgins1990Table5)
anova(m)
```

    ## Analysis of Variance of Aligned Rank Transformed Data
    ## 
    ## Table Type: Repeated Measures Analysis of Variance Table (Type I) 
    ## Model: Repeated Measures (aov)
    ## Response: art(DryMatter)
    ## 
    ##                       Error Df Df.res F value     Pr(>F)    
    ## 1 Moisture             Tray  3      8  23.833 0.00024199 ***
    ## 2 Fertilizer          Withn  3     24 122.402 1.1124e-14 ***
    ## 3 Moisture:Fertilizer Withn  9     24   5.118 0.00064665 ***
    ## ---
    ## Signif. codes:   0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Contrast tests
--------------

For an example of how to run contrast tests on an `art` model, see this vignette:

``` r
vignette("art-contrasts")
```

This vignette is also available [here](https://cran.r-project.org/web/packages/ARTool/vignettes/art-contrasts.html).

Problems
--------

Should you encounter any issues with this package, contact Matthew Kay (<mjskay@umich.edu>). If you have found a bug, please file it [here](https://github.com/mjskay/ARTool/issues/new) with minimal code to reproduce the issue.

Citations
---------

Kay M and Wobbrock J (2016). *ARTool: Aligned Rank Transform for Nonparametric Factorial ANOVAs*. R package version 0.10.4, <https://github.com/mjskay/ARTool>. DOI: [10.5281/zenodo.48543](http://dx.doi.org/10.5281/zenodo.48543).

Wobbrock J, Findlater L, Gergle D and Higgins J (2011). "The Aligned Rank Transform for Nonparametric Factorial Analyses Using Only ANOVA Procedures." In *Proceedings of the ACM Conference on Human Factors in Computing Systems (CHI '11)*, Vancouver, British Columbia (May 7-12, 2011). New York: ACM Press, pp. 143-146. <http://depts.washington.edu/aimgroup/proj/art/>. DOI: [10.1145/1978942.1978963](http://dx.doi.org/10.1145/1978942.1978963).


NEWS
ARTool 0.10.4
=============

Minor changes:

* Reference phia vignette using function call instead of non-canonical URL
* Cautionary note about standardized effect sizes in effect size vignette
* Added testInteractions example as alternative in contrasts vignette

Bugfixes:

* Dependency fixes for failed test at testthat/test.artlm.R:35 (our use of lsmeans in that
test requires some additional packages only declared as "suggests" in lsmeans, so now
we "suggest" them as well).


ARTool 0.10.1
=============

New features:

* New vignette describing effect size estimates
* Using lsmeans interactions argument instead of phia for interaction contrasts vignette

Bugfixes:

* Require R >= 3.2 and lsmeans >= 2.22 to fix some bugs in earlier versions


ARTool 0.10.0
=============

New features:

* Support for Error() terms in model formulas (resulting models are run using aov())
* Checks for numeric variables passed into formulas that may cause incorrect results (if the user intended data to be treated as categorical)
* More detailed ANOVA tables
* New vignette describing contrast tests, particularly for interactions (vignette("art-contrasts"))

Bugfixes:

* Formulas now correctly support arbitrary expressions as terms (rather than just column names).


ARTool 0.9.5
============

Testing fix for changes in upcoming version of lsmeans: round lsmeans p value tests to 5 decimal places to accommodate changes to Tukey adjustment


ARTool 0.9.4
============

Minor changes to testing based on updated version of testthat (0.10.0):

* More closely follow recommended testthat usage
* Skip tests requiring lsmeans if it is not installed


R/anova.art.R
# ANOVAs for art objects
# 
# Author: mjskay
###############################################################################


#' Aligned Rank Transform Analysis of Variance
#' 
#' Conduct analyses of variance on aligned rank transformed data.
#' 
#' This function runs several ANOVAs: one for each fixed effect term in the
#' model \code{object}. In each ANOVA, the independent variables are the same,
#' but the response is aligned by a different fixed effect term (if response is
#' "aligned") or aligned and ranked by that fixed effect term (if response is
#' "art"). These models are generated using \code{\link{artlm}}.
#' 
#' From each model, only the relevant output rows are kept (unless
#' \code{all.rows} is \code{TRUE}, in which case all rows are kept).
#' 
#' When \code{response} is \code{"art"} (the default), only one row is kept
#' from each ANOVA: the row corresponding to fixed effect term the response was
#' aligned and ranked by. These results represent nonparametric tests of
#' significance for the effect of each term on the original response variable.
#' 
#' When \code{response} is \code{"aligned"}, all rows \emph{except} the row
#' corresponding to the fixed effect term the response was aligned by are kept.
#' If the ART procedure is appropriate for this data, these tests should have
#' all effects "stripped out", and have an F value of ~0. If that is not the
#' case, another analysis should be considered. This diagnostic is tested by
#' \code{\link{summary.art}} and a warning generated if the F values are not
#' all approximately 0.
#' 
#' @name anova.art
#' @rdname anova.art
#' @aliases anova.art print.anova.art
#' @param object An object of class \code{\link{art}}.
#' @param response Which response to run the ANOVA on: the aligned responses
#' (\code{"aligned"}) or the aligned and ranked responses (\code{"art"}). This
#' argument is passed to \code{\link{artlm}}. See 'Details'.
#' @param type Type of ANOVAs to conduct. If \code{type} is \code{1} or
#' \code{"I"}, then conducts Type I ANOVAs using \code{\link{anova}}.
#' Otherwise, conducts Type II or Type III ANOVAs using \code{\link{Anova}}.
#' @param factor.contrasts The name of the contrast-generating function to be
#' applied by default to fixed effect factors. See the first element of
#' \code{\link{options}("contrasts")}. The default is to use
#' \code{"contr.sum"}, i.e. sum-to-zero contrasts, which is appropriate for
#' Type III ANOVAs (also the default). This argument is passed to
#' \code{\link{artlm}}.
#' @param test Test statistic to use. Default \code{"F"}. Note that some models
#' and ANOVA types may not support \code{"Chisq"}.
#' @param all.rows Show all rows of the resulting ANOVA tables? By default
#' (\code{FALSE}), shows only the rows that are relevant depending on the type
#' of \code{response}.
#' @param x An object of class \code{\link{art}}.
#' @param verbose When \code{TRUE}, sums of squares and reisdual sum of squares
#' in addition to degrees of freedom are printed in some Anova types (e.g.
#' repeated measures Anovas). Default \code{FALSE}, for brevity.
#' @param digits Digits of output in printed table; see \code{\link{print}}.
#' @param \dots Additional arguments passed to \code{\link{Anova}} or
#' \code{\link{anova}} by \code{anova.art} or to \code{\link{print}} by
#' \code{print.anova.art}.
#' @return An object of class \code{"anova"}, which usually is printed.
#' @author Matthew Kay
#' @seealso See \code{\link{art}} for an example. See also
#' \code{\link{summary.art}}, \code{\link{artlm}}.
#' @references Wobbrock, J. O., Findlater, L., Gergle, D., and Higgins, J. J.
#' (2011). The Aligned Rank Transform for nonparametric factorial analyses
#' using only ANOVA procedures. \emph{Proceedings of the ACM Conference on
#' Human Factors in Computing Systems (CHI '11)}.  Vancouver, British Columbia
#' (May 7-12, 2011). New York: ACM Press, pp. 143-146.
#' @keywords nonparametric
#' 
#' @export
anova.art = function(object, 
    response=c("art", "aligned"),
    type=c("III", "II", "I", 3, 2, 1), 
    factor.contrasts="contr.sum",
    test=c("F", "Chisq"),  
    all.rows=FALSE,
    ...
) {
    #sensible names for generic parameters
    m = object
    
    #match enum arguments
    response = match.arg(response)
    type = as.character(type)
    type = match.arg(type)
    test = match.arg(test)
    
    #get transformed responses based on response type requested
    responses = switch(response, 
        aligned=m$aligned, 
        art=m$aligned.ranks)
    
    #determine anova type to use
    type = if (type %in% c(1,"I")) "I"
    else if (type %in% c(2,"II")) "II"
    else if (type %in% c(3,"III")) "III"
        
    #are we going to need to show the term we aligned by
    #for each row of the output?
    show.aligned.by.term = response == "aligned" || all.rows
    
    #run linear models and anovas
    df = m$data
    anovas = NULL
    table.description = "" 
    for (j in 1:ncol(responses)) {	#can't use ldply here because it appears to drop row names when binding rows
        aligned.by.term = colnames(responses)[[j]]
        
        #get linear model
        m.l = artlm(m, aligned.by.term, response=response, factor.contrasts=factor.contrasts)
        
        #run anova and extract desired results 
        anova.j = flat.anova(m.l, type=type, test=test, ...)
        if (j == 1) table.description = attr(anova.j, "description")
        
        #extract desired result rows from anova
        #for art, this is the one row correponding to the effect we aligned and ranked by
        #for aligned, this is every effect *except* the one we aligned and ranked by 
        if (!all.rows) {
            include.row = anova.j$Term == aligned.by.term
            if (response == "aligned") include.row = !include.row
            anova.j = anova.j[include.row,]
        }
        
        #Add "Aligned By" column when needed to disambiguate
        if (nrow(anova.j) > 0) {	
            #if only one fixed effect we can get no rows here, e.g. if response="aligned" 
            #and all.rows=FALSE, so the above guard is necessary
            if (show.aligned.by.term) {
                anova.j = cbind(anova.j[,1,drop=FALSE], `Aligned By`=aligned.by.term, anova.j[,-1,drop=FALSE])
            }
        }
        
        anovas = rbind(anovas, anova.j)
    }
    
    #fill in the rest of the anova table metadata and return
    class(anovas) = c("anova.art", "anova", "data.frame")
    attr(anovas, "model") = 
        if (m$n.grouping.terms > 0) "lmer" 
        else if (m$n.error.terms > 0) "aov" 
        else "lm"
    attr(anovas, "table.description") = table.description
    attr(anovas, "response") = response
    attr(anovas, "response.term") = colnames(m$cell.means)[1]
    anovas	
}

### Generate p stars for a vector of p values
#' @importFrom stats symnum
p.stars = function(p.values) {
    unclass(symnum(p.values, corr = FALSE, na = FALSE, cutpoints = c(0, 
                            0.001, 0.01, 0.05, 0.1, 1), symbols = c("***", "**", 
                            "*", ".", " ")))
}

#' @rdname anova.art
#' @importFrom magrittr %<>%
#' @export
print.anova.art = function(x, verbose=FALSE, digits=5, ...) {
    #print heading and metadata
    cat("Analysis of Variance of Aligned Rank Transformed Data\n\n")
    cat("Table Type:", attr(x, "table.description"), "\n")
    cat("Model:", switch(attr(x, "model"),
        lm = "No Repeated Measures (lm)\n",
        aov = "Repeated Measures (aov)\n",
        lmer = "Mixed Effects (lmer)\n",
    ))
    cat(sep="", "Response: ", attr(x, "response"), "(", attr(x, "response.term"), ")\n\n")
    
    #format p values
    p.col = last(which(grepl("^(P|Pr)\\(", names(x))))
    stars.legend = if (!is.na(p.col)) {
        #add column for p stars
        stars = p.stars(x[[p.col]])
        x = cbind(x, ` ` = stars)
        #reformat p values for printing
        x[[p.col]] %<>% format.pval()
        #return stars legend
        attr(stars, "legend")
    } else NULL

    #generate row names from Terms
    rownames(x) = paste(1:nrow(x), x$Term)
    x$Term = NULL

    #abbreviate columns
    if (!is.null(x$Error)) x$Error %<>% abbreviate(5)
    if (!is.null(x$`Aligned By`)) x$`Aligned By` %<>% abbreviate(10)
    
    #drop "Sum Sq" (etc) columns when not doing verbose output 
    if (!verbose) x %<>% select(everything(), -contains("Sum Sq"), -contains("Mean Sq"))
    
    #print table
    print.data.frame(x, digits=digits, ...)
    
    #print legend
    if (!is.null(stars.legend)) {
        cat("---\nSignif. codes:  ", stars.legend, "\n")
    }
}
